lib/puppet/type/vault_cert.rb

```diff
@@ -0,0 +1,170 @@
+      unless File.directory?(value)
+        raise ArgumentError, "Directory not found for: #{value}"
+      end
+    end
+  end
+
+  newparam(:priv_key_name) do
+    desc 'The name of the private key'
+    defaultto do
+      cert_name = @resource[:cert_name]
+      extension = File.extname(cert_name)
+      File.basename(cert_name, extension) + '.key'
+    end
+  end
+
+  newparam(:priv_key_dir) do
+    desc 'The directory that the private key lives in'
+    defaultto do
+      Pathname.new(@resource[:cert_dir])
+    end
+    validate do |value|
+      path = Pathname.new(value)
+      # Verify that an absolute path was given
+      unless path.absolute?
+        raise ArgumentError, "Path must be absolute: #{path}"
+      end
+      # Verify that the given directory exists
+      unless File.directory?(value)
+        raise ArgumentError, "Directory not found for: #{value}"
+      end
+    end
+  end
+
+  newparam(:new_cert_dir) do
+    desc 'The path to save the new certificate in'
+    # Default to the directory of the given cert
+    defaultto do
+      @resource[:cert_dir]
+    end
+    validate do |value|
+      path = Pathname.new(value)
+      # Verify that an absolute path was given
+      unless path.absolute?
+        raise ArgumentError, "Path must be absolute: #{path}"
+      end
+      # Verify that the given directory exist
+      unless File.directory?(value)
+        raise ArgumentError, "Directory not found for: #{value}"
+      end
+    end
+  end
+
+  newparam(:new_priv_key_dir) do
+    desc 'The path to save the new certificate in'
+    # Default to the directory of the given private key
+    defaultto do
+      @resource[:priv_key_dir]
+    end
+    validate do |value|
+      path = Pathname.new(value)
+      # Verify that an absolute path was given
+      unless path.absolute?
+        raise ArgumentError, "Path must be absolute: #{path}"
+      end
+      # Verify that the given directory exist
+      unless File.directory?(value)
+        raise ArgumentError, "Directory not found for: #{value}"
+      end
+    end
+  end
+
+  newparam(:key_password) do
+    desc 'The optional password for the private key'
+  end
+
+  newparam(:auth_type) do
+    desc 'authentication type of the private key'
+    defaultto('ec')
+    validate do |value|
+      acceptable_values = ['dsa', 'rsa', 'ec']
+      unless acceptable_values.include? value.downcase
+        raise ArgumentError, "auth_type must be one of: #{acceptable_values.join(', ')}"
+      end
+    end
+  end
+
+  newparam(:regenerate_ttl) do
+    desc 'Re-generate and replace the certificate this many hours before it expires.'
+    defaultto(3)
+  end
+
+  newparam(:cert_ttl) do
+    desc 'TTL to give the new cert'
+    defaultto('720h')
+  end
+
+  newparam(:owner) do
+    desc 'Owner to assign for the new cert and key'
+    defaultto('root')
+  end
+
+  newparam(:group) do
+    desc 'Group to assign for the new cert and key'
+    defaultto('root')
```

Comment:
@jschoewe need to add an `alt_sans` parameter here too for named subject-alternative-names that are not IP addresses, example (myservice.domain.tld,myservice)